src/server/snap.rs

```diff
@@ -189,10 +174,123 @@ fn send_snap(
     Ok(send)
 }
 
+struct RecvSnapContext {
+    key: SnapKey,
+    file: Option<Box<Snapshot>>,
+    raft_msg: RaftMessage,
+    finished: bool,
+}
+
+fn recv_snap<R: RaftStoreRouter + 'static>(
+    task: Receiver<Option<SnapshotChunk>>,
+    snap_mgr: SnapManager,
+    raft_router: R,
+) -> impl Future<Item = (), Error = ()> {
+    let snap_mgr_1 = snap_mgr.clone();
+    let all_chunks_success = task.into_future().map_err(|_| ()).and_then(
+        move |(head, chunks)| -> Box<Future<Item = RecvSnapContext, Error = ()> + Send> {
+            // head is None means we meet gRPC error when receiving messages.
+            let head = match head {
+                Some(h) => h,
+                None => return box future::err(()),
+            };
+            let mut context = match get_context_from_head_chunk(head, &snap_mgr_1) {
+                Ok(context) => context,
+                Err(_) => return box future::err(()),
+            };
+            if context.file.is_none() {
+                context.finished = true;
+                return box future::ok(context);
+            }
+
+            snap_mgr_1.register(context.key.clone(), SnapEntry::Receiving);
+            box chunks.fold(context, |mut context, chunk| {
+                let mut chunk = match chunk {
+                    Some(c) => c,
+                    None => {
+                        context.finished = true;
+                        return Ok(context);
+                    }
+                };
+
+                let data = chunk.take_data();
+                if !data.is_empty() {
+                    let (key, file) = (&context.key, context.file.as_mut().unwrap());
+                    if let Err(e) = file.write_all(&data) {
+                        let path = file.path();
+                        error!("{} failed to write snapshot file {}: {}", key, path, e);
+                        return Err(());
```

Comment:
if return error, where to deregister the snapshot key?

Reply:
can you use the same place to handle error? I worry we may forget to deregister later.

Reply:
Nice catch. Fixed.